src/test/groovy/pl/futurecollars/invoicing/controllers/InvoiceControllerExhaustiveTest.groovy

```diff
@@ -0,0 +1,236 @@
+        invoice1.setId(invoices[1].getId())
+
+        then:
+        invoices.length == 2
+        invoices[0] == invoice
+        invoices[1] == invoice1
+    }
+
+    def "should update not existing invoice"() {
+        given:
+        def invoiceAsJson = jsonService.convertToJson(invoice)
+
+        when:
+        def postResponse = mockMvc.perform(
+                post("/invoices").content(invoiceAsJson).contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        def id = jsonService.convertToObject(postResponse, Invoice.class).getId()
+        invoice.setId(id)
+
+        UUID updatedId
+        do updatedId = UUID.randomUUID()
+        while (updatedId == id)
+
+        def updatedInvoice = InvoiceFixture.invoice(1)
+        updatedInvoice.setId(updatedId)
+        def updatedInvoiceAsJson = jsonService.convertToJson(updatedInvoice)
+
+        def putResponse = mockMvc.perform(
+                put("/invoices").content(updatedInvoiceAsJson).contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        def response = mockMvc.perform(get("/invoices"))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        def invoices = jsonListService.convertToObject(response, Invoice[].class)
+
+        then:
+        invoice != updatedInvoice
+        invoices[0] == invoice
+        putResponse == ""
+    }
+
+    def "should delete not existing invoice"() {
+        given:
+        def invoiceAsJson = jsonService.convertToJson(invoice)
+
+        when:
+        def postResponse = mockMvc.perform(
+                post("/invoices").content(invoiceAsJson).contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        def id = jsonService.convertToObject(postResponse, Invoice.class).getId()
+        invoice.setId(id)
+
+        UUID updatedId
+        do updatedId = UUID.randomUUID()
+        while (updatedId == id)
+
+        def deleteResponse = mockMvc.perform(
+                delete("/invoices/" + updatedId))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        def response = mockMvc.perform(get("/invoices"))
+                .andExpect(status().isOk())
+                .andReturn()
+                .response
+                .contentAsString
+
+        def invoices = jsonListService.convertToObject(response, Invoice[].class)
+
+        then:
+        deleteResponse == "false"
+        invoices[0] == invoice
+    }
+
+    def "should delete second of 3 invoices"() {
```

Comment:
I would add helper methods to add invoices to database to reduce repetition and to make tests more concise